src/osd/MissingLoc.cc

```diff
@@ -187,49 +187,30 @@ void MissingLoc::check_recovery_sources(const OSDMapRef& osdmap)
 
 void MissingLoc::remove_stray_recovery_sources(pg_shard_t stray)
 {
-  set<pg_shard_t> now_stray;
-  for (set<pg_shard_t>::iterator p = missing_loc_sources.begin();
-       p != missing_loc_sources.end();
-       ) {
-    if (*p != stray) {
-      ++p;
-      continue;
-    }
-    ldout(cct, 10) << __func__ << " source osd." << *p << " now stray" << dendl;
-    now_stray.insert(*p);
-    missing_loc_sources.erase(p++);
```

Comment:
I wonder if you should still remove _stray_ from _missing_loc_sources_ if present.  The only thing you care about is if it isn't in _missing_loc_source_ still remove it from _missing_loc_.

Or should any code that adds something to _missing_loc_ insert into _missing_loc_sources_?  Is this bug really that the repair code doesn't fix _missing_loc_sources_.  That would be PrimaryLogPG::primary_error() calling PeeringState::force_object_missing() and MissingLoc::rebuild() which doesn't affect _missing_loc_sources_.

Reply:
We could still remove stray from missing_loc_sources to make it consistent, but we only consult missing_loc when looking for correct copies during recovery, so this should also work. There are only two places where we insert items into missing_loc_sources, so it is possible this statement is not true: "any code that adds something to missing_loc insert into missing_loc_sources"

Reply:
@dzafman added the bit about cleaning up missing_loc_sources